File: leetcode/pow_function/Pow.cpp
#include <iostream> 

using namespace std; 

class Solution {
public:
    double pow(double x, int n) { 
    	long long exp = n; 
    	 	
        if (exp == 0) {
            return 1; 
        }
        
        if (exp < 0) {
            x = 1 / x; 
            exp = exp * (-1); 
        }
        
        return powHelper(x, exp); 
    }
    
    double powHelper(double base, long long n) {
        if (n == 0) {
            return 1; 
        }
        
        if (n == 1) {
            return base; 
        }
        
        double divideTwo = powHelper(base, n / 2); 
        double modTwo = powHelper(base, n % 2);
         
        return  divideTwo * divideTwo * modTwo; 
    }
};

int main() {
	Solution sln; 
	cout << sln.pow(0.00001, 2147483647) << endl; 
	return 0; 
}

File: leetcode4/Pow.java
public class Solution {
    public double pow(double x, int n) {
        if (n == 0) {
            return 1.0; 
        }
        
        boolean isNeg = false; 
        
        if (n == 1) {
            return x; 
        }
        
        if (n < 0) {
            isNeg = true; 
            n = -n; 
        }
        
        double temp1 = pow(x, n / 2);
        double temp2 = pow(x, n % 2);
        
        if (!isNeg) {
            return temp1 * temp1 * temp2; 
        } else {
            return 1 / (temp1 * temp1 * temp2); 
        }
    }
}
